src/Symfony/Component/Security/Http/Authenticator/FormLoginAuthenticator.php

```diff
@@ -0,0 +1,143 @@
+use Symfony\Component\Security\Http\Authenticator\Passport\Badge\RememberMeBadge;
+use Symfony\Component\Security\Http\Authenticator\Passport\Credentials\PasswordCredentials;
+use Symfony\Component\Security\Http\Authenticator\Passport\Passport;
+use Symfony\Component\Security\Http\Authenticator\Passport\PassportInterface;
+use Symfony\Component\Security\Http\HttpUtils;
+use Symfony\Component\Security\Http\ParameterBagUtils;
+
+/**
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ * @author Fabien Potencier <fabien@symfony.com>
+ *
+ * @final
+ * @experimental in 5.1
+ */
+class FormLoginAuthenticator extends AbstractLoginFormAuthenticator
+{
+    private $httpUtils;
+    private $userProvider;
+    private $successHandler;
+    private $failureHandler;
+    private $options;
+
+    public function __construct(HttpUtils $httpUtils, UserProviderInterface $userProvider, AuthenticationSuccessHandlerInterface $successHandler, AuthenticationFailureHandlerInterface $failureHandler, array $options)
+    {
+        $this->httpUtils = $httpUtils;
+        $this->userProvider = $userProvider;
+        $this->successHandler = $successHandler;
+        $this->failureHandler = $failureHandler;
+        $this->options = array_merge([
+            'username_parameter' => '_username',
+            'password_parameter' => '_password',
+            'check_path' => '/login_check',
+            'post_only' => true,
+
+            'enable_csrf' => false,
+            'csrf_parameter' => '_csrf_token',
+            'csrf_token_id' => 'authenticate',
+        ], $options);
+    }
+
+    protected function getLoginUrl(Request $request): string
+    {
+        return $this->httpUtils->generateUri($request, $this->options['login_path']);
+    }
+
+    public function supports(Request $request): bool
+    {
+        return ($this->options['post_only'] ? $request->isMethod('POST') : true)
+            && $this->httpUtils->checkRequestPath($request, $this->options['check_path']);
+    }
+
+    public function authenticate(Request $request): PassportInterface
+    {
+        $credentials = $this->getCredentials($request);
+        $user = $this->userProvider->loadUserByUsername($credentials['username']);
+        if (!$user instanceof UserInterface) {
+            throw new AuthenticationServiceException('The user provider must return a UserInterface object.');
+        }
+
+        $badges = [new RememberMeBadge()];
+        if ($this->options['enable_csrf']) {
+            $badges[] = new CsrfTokenBadge($this->options['csrf_token_id'], $credentials['csrf_token']);
+        }
+
+        if ($this->userProvider instanceof PasswordUpgraderInterface) {
+            $badges[] = new PasswordUpgradeBadge($credentials['password'], $this->userProvider);
+        }
+
+        return new Passport($user, new PasswordCredentials($credentials['password']), $badges);
+    }
+
+    public function createAuthenticatedToken(PassportInterface $passport, $providerKey): TokenInterface
```

Comment:
We should probably at least have phpdoc that says that this will actually be a `Passport` object so that the `->getUser()` method exists.